File: data-jpa/src/main/java/top/bettercode/summer/data/jpa/query/DefaultSpecMatcher.java
package top.bettercode.summer.data.jpa.query;

/**
 * @author Peter Wu
 */
public class DefaultSpecMatcher<T> extends SpecMatcher<T, DefaultSpecMatcher<T>> {

  private static final long serialVersionUID = 1L;

  protected DefaultSpecMatcher(SpecMatcherMode matcherMode, T probe) {
    super(matcherMode, probe);
  }

  /**
   * 创建 SpecMatcher 实例
   *
   * @param <T> T
   * @return SpecMatcher 实例
   */
  public static <T> SpecMatcher<T, DefaultSpecMatcher<T>> matching() {
    return new DefaultSpecMatcher<>(SpecMatcherMode.ALL, null);
  }

  public static <T> SpecMatcher<T, DefaultSpecMatcher<T>> matching(T probe) {
    return new DefaultSpecMatcher<>(SpecMatcherMode.ALL, probe);
  }

  /**
   * 创建 SpecMatcher 实例
   *
   * @param <T> T
   * @return SpecMatcher 实例
   */
  public static <T> SpecMatcher<T, DefaultSpecMatcher<T>> matchingAny() {
    return new DefaultSpecMatcher<>(SpecMatcherMode.ANY, null);
  }

  public static <T> SpecMatcher<T, DefaultSpecMatcher<T>> matchingAny(T probe) {
    return new DefaultSpecMatcher<>(SpecMatcherMode.ANY, probe);
  }

}

File: data-jpa/src/main/kotlin/top/bettercode/summer/data/jpa/query/DefaultSpecMatcher.kt
package top.bettercode.summer.data.jpa.query

/**
 * @author Peter Wu
 */
open class DefaultSpecMatcher<T : Any?> protected constructor(matcherMode: SpecMatcherMode, probe: T?) : SpecMatcher<T, DefaultSpecMatcher<T>>(matcherMode, probe) {
    companion object {

        /**
         * 创建 SpecMatcher 实例
         *
         * @param <T> T
         * @return SpecMatcher 实例
        </T> */
        fun <T : Any?> matching(): SpecMatcher<T, DefaultSpecMatcher<T>> {
            return DefaultSpecMatcher(SpecMatcherMode.ALL, null)
        }

        fun <T : Any?> matching(probe: T): SpecMatcher<T, DefaultSpecMatcher<T>> {
            return DefaultSpecMatcher(SpecMatcherMode.ALL, probe)
        }

        /**
         * 创建 SpecMatcher 实例
         *
         * @param <T> T
         * @return SpecMatcher 实例
        </T> */
        fun <T : Any?> matchingAny(): SpecMatcher<T, DefaultSpecMatcher<T>> {
            return DefaultSpecMatcher(SpecMatcherMode.ANY, null)
        }

        fun <T : Any?> matchingAny(probe: T): SpecMatcher<T, DefaultSpecMatcher<T>> {
            return DefaultSpecMatcher(SpecMatcherMode.ANY, probe)
        }
    }
}
